### ISSUE TITLE
mqtt consumer doesn't receive mesg from pulsar-mop

### ISSUE THREAD

**Issue by rjiang-ptm** [streamnative/mop#20]
### Problem Statement

- pulsar-standalone 2.5.2
- installed mop and configured as instructed
- currently setup in a standalone mode in both docker container and cluster
- used different clients to publish/subscribe to topic

case 1:
used _mqtt cli_ directly to test producer/consumer:
```
mqtt test -h localhost -p 1883
MQTT 3: OK
	- Maximum topic length: 0 bytes
	- QoS 0: Received 0/10 publishes in 10033.91ms
	- QoS 1: Received 0/10 publishes in 10156.59ms
	- QoS 2: 

```
case 2:
used _patho_ python clients to test producer/consumer, also cannot receive mesg as consumer

case 3:
mqtt producer sends to topic (on 1883) and regular pulsar consumer (on 6650) received the message. However, messages sent out to the topic that mqtt consumer is supposed to receive didn't happen. That means producer seems to work fine but consumer was not able to receive messages.

In both cases, there was no errors in pulsar broker logs. The connection/publish/subscribe seem all succeeded but no message received on consumer end. We tried both qos = 0 and qos = 1, same effect.

What could be the issue here?

**Comment by rjiang-ptm** [comment]
Pls let me know if you need any more detailed info...

**Comment by wolfstudy** [comment]
@rjiang-ptm Sorry for the reply later. This seems to be a compatibility issue. I tried to use the 2.6.1 version, and I encountered the same error as you. But 2.6.0 can work normally, this is because the client of 2.6.0 is used inside mop, we may need to confirm the specific incompatibility.

![image](https://user-images.githubusercontent.com/20965307/96233029-b76b3900-0fca-11eb-922c-35cc18b20b98.png)

**Comment by rjiang-ptm** [comment]
Thank you for the reply!

Based on my current testing, the 2.6.0 still doesn’t receive any mesg published. It seems the mqtt client connects to the broker and sends the subscription properly:

```
process CONNECT message. CId=kep1016, username=null
18:31:18.253 [pulsar-io-51-1] INFO  io.streamnative.pulsar.handlers.mqtt.support.ProtocolMethodProcessorImpl - Configuring connection. CId=kep1016
18:31:18.254 [pulsar-io-51-1] INFO  io.streamnative.pulsar.handlers.mqtt.support.ProtocolMethodProcessorImpl - Sending connect ACK. CId=kep1016
18:31:18.256 [pulsar-io-51-1] INFO  io.streamnative.pulsar.handlers.mqtt.support.ProtocolMethodProcessorImpl - The connect ACK has been sent. CId=kep1016
18:31:18.257 [pulsar-io-51-1] INFO  io.streamnative.pulsar.handlers.mqtt.support.ProtocolMethodProcessorImpl - The CONNECT message has been processed. CId=kep1016, username=null success=true
18:31:18.259 [pulsar-io-51-1] INFO  io.streamnative.pulsar.handlers.mqtt.support.ProtocolMethodProcessorImpl - [Subscribe] [[id: 0xb27316bf, L:/10.1.1.50:1883 - R:/10.1.1.1:34216]] msg: MqttSubscribeMessage[fixedHeader=MqttFixedHeader[messageType=SUBSCRIBE, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=11], variableHeader=MqttMessageIdVariableHeader[messageId=1], payload=MqttSubscribePayload[MqttTopicSubscription[topicFilter=tp1016, qualityOfService=AT_LEAST_ONCE]]]

```

After the mqtt consumer sends out the subscription, there is no SUBACK or message received.

Which mqtt client are you using? What specifics do you think we should fine tune to make this work on 2.6.0?

Thanks!

**Comment by wolfstudy** [comment]
> Which mqtt client are you using? What specifics do you think we should fine tune to make this work on 2.6.0?

I am using the latest version of `mqtt-box`.

**Comment by rjiang-ptm** [comment]
Is there any intricate details that other mqtt clients don't comply and only this mqtt-box works for pulsar-mop? Can you specify what needs to be done on the mqtt consumer side to ensure it can receive the message on 1883? As mentioned above, we used hivemq (mqtt cli) and patho mqtt client, both couldn't receive messages on 1883. Our pulsar broker (2.6.0) with mop seems to be good as I was able to test with mqtt-box (it does work on 2.6.0). 

So the question becomes what needs to be done to ensure other pulsar consumer (mqtt clients) can work as welll?

**Comment by rjiang-ptm** [comment]
Confirmed that subscription name is required for pulsar consumer. Given the clients we used (open-source) mostly don't have subscription name as an option (only topic), the consumer on those clients failed to receive the SUBACK and published topic mesg. This could be the root cause for our consumer not receiving mesg.

Suggestion from the pulsar slack channel:

> yes, you always need a subscription name, but sometimes that subscription name can be automatically generated. It might be that MoP should be able to handle no subscription and just use a generated name (probably like the client hostname or something)

**Comment by vicaya** [comment]
@rjiang-ptm, does commit #22 fix the issue?

**Comment by rjiang-ptm** [comment]
> @rjiang-ptm, does commit #22 fix the issue?

Pls refer to my comments above, the issue could be related to subscription name above, does #22 address that?

**Comment by codelipenghui** [comment]
Seems we need to create an internal subscription if the mqtt-client do not provide a sub name.

**Comment by codelipenghui** [comment]
@rjiang-ptm Could you please provide a client test demo? so that we can ensure this fix can works for you. Thanks

**Comment by rjiang-ptm** [comment]
I'm using paho python client: https://pypi.org/project/paho-mqtt/

Here is a sample consumer code in paho:

`
import paho.mqtt.client as mqtt
import paho.mqtt.subscribe as subscribe

ip = "your ip"
port = 31883
suf = "1016"
# tp = "tp" + suf
tp = "persistent://public/default/tp" + suf

# subscribe with callback
# The callback for when the client receives a CONNACK response from the server.
print("MQTT consumer..." + suf)
def on_connect(client, userdata, flags, rc):
    print("on_connect(): client connected...")
    # client.subscribe(topic, 1)
    # print("suscribed to " + topic)

# The callback for when a PUBLISH message is received from the server.
def on_message(client, userdata, msg):
    print("on_message(): {}".format(msg))
    print(msg.topic + ": " + str(message.payload.decode("utf-8")))

def on_subscribe(client, userdata, mid, granted_qos):
    print("on_subscribe(): mid = {}, qos = {}".format(mid, granted_qos))

def on_log(client, userdata, level, buf):
    print("on_log: ",buf)

client = mqtt.Client(client_id="kep"+suf, clean_session=False)
print("got the mqtt client")

client.on_log = on_log
client.on_connect = on_connect
client.on_subscribe = on_subscribe
client.on_message = on_message

client.connect(ip, port, 60)
client.subscribe(tp, 0)
print("suscribe() called to " + tp)

# Blocking call that processes network traffic, dispatches callbacks and
# handles reconnecting.
# Other loop*() functions are available that give a threaded interface and a
# manual interface.
client.loop_forever()
`

**Comment by wolfstudy** [comment]
hello @rjiang-ptm 

It seems that currently for the specified client id, there is no command to return SUBACK when subscribing to the topic repeatedly. We should fix this problem. If you want to use it now, I have made a simple modification to the code you provided. You can use this code to test, and my multiple tests are OK.

```
import paho.mqtt.client as mqtt
import paho.mqtt.subscribe as subscribe

ip = "127.0.0.1"
port = 1883
suf = "1017"
# tp = "tp" + suf
tp = "persistent://public/default/tp" + suf

# subscribe with callback
# The callback for when the client receives a CONNACK response from the server.
print("MQTT consumer..." + suf)
def on_connect(client, userdata, flags, rc):
    print("on_connect(): client connected...")
    # client.subscribe(topic, 1)
    # print("suscribed to " + topic)

# The callback for when a PUBLISH message is received from the server.
def on_message(client, userdata, msg):
    print("on_message(): {}".format(msg))
    print(msg.topic + ": " + str(msg.payload.decode("utf-8")))

def on_subscribe(client, userdata, mid, granted_qos):
    print("on_subscribe(): mid = {}, qos = {}".format(mid, granted_qos))

def on_log(client, userdata, level, buf):
    print("on_log: ",buf)

client = mqtt.Client()
print("got the mqtt client")

client.on_log = on_log
client.on_connect = on_connect
client.on_subscribe = on_subscribe
client.on_message = on_message

client.connect(ip, port, 60)
client.subscribe(tp, 0)
print("suscribe() called to " + tp)

# Blocking call that processes network traffic, dispatches callbacks and
# handles reconnecting.
client.loop_forever()
```

**Comment by rjiang-ptm** [comment]
Could you explain a bit what you changed? And are we talking about the same issue? The issue raised here was related to missing subscription name from mqtt clients (often the case) while subscription name is required by pulsar.

Pls note this is just a sample client, we won't really use this code for our system as we'll use some other mqtt client in production. Fundamentally if you think missing subscription name (in many mqtt clients) is an issue, then should mop handler handle such case? 

Again, does this suggestion make sense -

Suggestion from the pulsar slack channel:

> yes, you always need a subscription name, but sometimes that subscription name can be automatically generated. It might be that MoP should be able to handle no subscription and just use a generated name (probably like the client hostname or something)

In other words, in case of mqtt, if the clients don't supply a subscription name (which seems to be the case), the mop handler could auto-generate a subname for the consumer so pulsar can properly handle it? Or you could suggest other ways to go around the problem?

**Comment by wolfstudy** [comment]
> Could you explain a bit what you changed?

1. replace `message` with `msg`

```
print(msg.topic + ": " + str(message.payload.decode("utf-8")))
```

Because you did not define the message field, you only defined msg, right?

2. remove client_id from the demo code

```
client = mqtt.Client(client_id="kep"+suf, clean_session=False)
print("got the mqtt client")
```

> And are we talking about the same issue? The issue raised here was related to missing subscription name from mqtt clients (often the case) while subscription name is required by pulsar.

Sure, pulsar client needs to subscribe to the name, this is no problem. In mop, we use client_id as the name of the subscription.

> Pls note this is just a sample client, we won't really use this code for our system as we'll use some other mqtt client in production. Fundamentally if you think missing subscription name (in many mqtt clients) is an issue, then should mop handler handle such case?

Again, In the processing of mop, client_id is used as the name of the subscription, otherwise no client can receive the message correctly, right?

**Comment by wolfstudy** [comment]
Did you try to re-execute the test using my modified code?

**Comment by wolfstudy** [comment]
In the processing logic here, there is indeed a problem. When repeatedly subscribing to the same topic with the same client_id, the SUB_ACK command is not correctly returned. I think this is what we need to deal with and pay attention to.

**Comment by rjiang-ptm** [comment]
I see, so the client_id is used as the subname internally in mop. 

Going forward, we should not specify any client_id or options like clean_session config params?

**Comment by rjiang-ptm** [comment]
> Did you try to re-execute the test using my modified code?

Yes, the current temp solution is to remove the client_id param in the consumer:  `client = mqtt.Client()`

**Comment by vicaya** [comment]
@wolfstudy, so the main bug is that the current mop doesn't handle persistent session (clean_session=False) correctly? According to the mqtt spec, if client id is not specified, no broker state will be kept beyond the connection, which means QoS beyond 0 is not possible.

**Comment by wolfstudy** [comment]
> @wolfstudy, so the main bug is that the current mop doesn't handle persistent session (clean_session=False) correctly? According to the mqtt spec, if client id is not specified, no broker state will be kept beyond the connection, which means QoS beyond 0 is not possible.

Yes, It looks like this, we will fix this problem as soon as possible.
